### DISCUSSION TITLE
The reception window isn't opened at the right time (v4.4.3)

### DISCUSSION THREAD

**Discussion by FranciscoRV** [Lora-net/LoRaMac-node#943] [Ideas]
Hi, 
I have a problem with the windows delay duration, the reception window isn't opened at the right time.

I see that the end of a transmission is notified with **OnRadioTxDone**, LoRaMacRadioEvents.Events.TxDone is set to 1, this flag is processed inside LoRaMacProcess() function and this function calls **ProcessRadioTxDone** which starts the timers RxWindowTimerX. 

LoRaMacProcess is called in the main loop so if the end of a transmission **(between OnRadioTxDone and ProcessRadioTxDone)**  arrive during a microcontroller treatment the windows delay will depend on the duration of this treatment.

To avoid this we could move the TimerStart( &MacCtx.RxWindowTimerX ) in OnRadioTxDone? or move the ProcessRadioTxDone function under interrupt?

**Comment by Pirakutu** [comment]
Hi guys,

Did you guys have a look at this? It seems to be something crucial IMO.

Cheers!

**Reply by mluis1** [reply]
No, we did not had the time yet.

It is in our to do list to figure out the best way to implement this enhancement.

**Reply by mluis1** [reply]
In order to better understand the execution environment I would like to ask the following:
- Which radio is being used?
- Are you using an RTOS?

Depending on the radio the IRQ processing is different.
For SX127x radios the OnRadioTxDone is called while running the radio irq handler. (your proposition could indeed fix the observed issue)
For SX126x and LR1110 radios the OnRadioTxDone is called by the main process as shown in below code excerpt. (at the moment I think your solution would not fully fix the observed issue)
```c
void LmHandlerProcess( void )
{
    uint16_t size = 0;

    // Process Radio IRQ
    if( Radio.IrqProcess != NULL )
    {
        Radio.IrqProcess( );
    }

    // Processes the LoRaMac events
    LoRaMacProcess( );

...
}
```

While we don't find a proper fix which is OK for all supported radios I would recommend that when executing a lengthy process to call `Radio.IrqProcess` and `LoRaMacProcess` as often as possible inside the lengthy process function.

In case an RTOS is used I would recommend to create a high priority task that waits for a radio event and then executes `Radio.IrqProcess` and `LoRaMacProcess`. Like this, one ensures that the radio and mac processing are always executed as soon as possible without interruptions.

Please also note that we are currently working on updating the SX127x and SX126x radio drivers implementation in order for them to expose an API similar to the LR1110 radio.
This means that in the future all supported radios should behave in the same way.
